### PR TITLE
Add option for "out" as defaultConsent

### PR DESCRIPTION
**Pull request by jonsnyder**
## Description

Add a configuration option for defaultConsent as "out". When this option is configured, all events that occur before explicit consent are dropped.

## Related Issue

https://jira.corp.adobe.com/browse/PDCL-1241

## Motivation and Context

Some customers do not like the queuing mechanism that the "pending" option provides.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [x] I have made any necessary test changes and all tests pass.
- [x] I have run the Sandbox successfully.

### CODE REVIEW

File: src/core/consent/createConsentStateMachine.js [hunk 1]
```diff
@@ -31,30 +35,74 @@ export default () => {
...
+        .reject(createDeclinedConsentError("The user declined consent."));
     }
   };
 
+  const awaitInitial = () =>
+    Promise.reject(new Error("Consent has not been initialized."));
+  const awaitInDefault = () => Promise.resolve();
   const awaitIn = () => Promise.resolve();
-  const awaitOut = () => Promise.reject(createDeclinedConsentError());
+  const awaitOutDefault = () =>
```
Comment (jfkhoury) [inline src/core/consent/createConsentStateMachine.js:48]: I HAD to go suggest a different log message right?? 😄  🙄

**Review by jfkhoury — APPROVED** [review]
That was more work than I thought, awesome job @jonsnyder 🔑

File: src/core/consent/createConsentStateMachine.js [hunk 2]
```diff
@@ -14,14 +14,18 @@ import { defer } from "../../utils";
...
 export const DECLINED_CONSENT = "The user declined consent.";
 export const DECLINED_CONSENT_ERROR_CODE = "declinedConsent";
+export const CONSENT_SOURCE_DEFAULT = "default";
+export const CONSENT_SOURCE_INITIAL = "initial";
+export const CONSENT_SOURCE_NEW = "new";
 
-const createDeclinedConsentError = () => {
-  const error = new Error(DECLINED_CONSENT);
+const createDeclinedConsentError = errorMessage => {
+  const error = Error(errorMessage);
```
Comment (Aaronius) [inline src/core/consent/createConsentStateMachine.js:22]: Did you intentionally remove `new`?

**Review by Aaronius — APPROVED** [review]
Very nice! Thanks @jonsnyder!

File: src/core/consent/createConsentStateMachine.js (hunk 2 above)
Comment (Aaronius) [reply src/core/consent/createConsentStateMachine.js:22]: If this was intentional, let's talk, since we're using `new` everywhere else I believe.

Comment (jonsnyder) [reply src/core/consent/createConsentStateMachine.js:22]: Not intentional, I'll fix it.

File: src/core/consent/createConsentStateMachine.js [hunk 3]
```diff
@@ -31,30 +35,74 @@ export default () => {
...
   return {
-    in() {
-      runAll();
-      this.awaitConsent = awaitIn;
+    in(source) {
+      if (source === CONSENT_SOURCE_DEFAULT) {
+        this.awaitConsent = awaitInDefault;
+      } else {
+        if (source === CONSENT_SOURCE_INITIAL) {
+          logger.info("Loaded user consent preferences.");
```
[the remark this answers is not in the document]
Comment (jonsnyder) [reply src/core/consent/createConsentStateMachine.js:66]: Great suggestion. I do like declined consent better than opted out.

**Comment by Aaronius** [comment]
@jonsnyder your last commit looks good. Can we merge this?

**Pull request merged**